### PR TITLE
✨ add explicit securitycontexts to controllers

### PR DESCRIPTION
**Pull request by tuminoid**
Set explicit, secure securityContexts for the controller manager deployment and containers instead of relying on defaults or fallbacks.

Only actual change here is enabling runtimeDefault seccompPolicy, instead of running as Unconfined.
https://kubernetes.io/docs/tutorials/security/seccomp/

Also, reindent poorly indented command block.

### CODE REVIEW

**Comment by tuminoid** [comment]
Hello maintainers! I'd like to hear your opinion on adding explicit securityContexts to CAPI controller manager.

- Adding them to manifest vs leaving it for cluster admins and cluster/namespace wide policies?
- Directly to the manager manifest vs creating kustomization?
- Adding seccompProfile, changing unconfined -> runtimeDefault?-

**Comment by sbueringer** [comment]
> Hello maintainers! I'd like to hear your opinion on adding explicit securityContexts to CAPI controller manager.
> 
> * Adding them to manifest vs leaving it for cluster admins and cluster/namespace wide policies?
> * Directly to the manager manifest vs creating kustomization?
> * Adding seccompProfile, changing unconfined -> runtimeDefault?-

1. I would add them to our manifests like you did
2. I would add them to the manager as you did - because it's simpler
3. I'm fine with adding seccompProfile assuming it will work on all Kubernetes clusters (I'm not really familiar with the seccomp feature in Kubernetes)

**Comment by tuminoid** [comment]
> > Hello maintainers! I'd like to hear your opinion on adding explicit securityContexts to CAPI controller manager.
> > 
> > * Adding them to manifest vs leaving it for cluster admins and cluster/namespace wide policies?
> > * Directly to the manager manifest vs creating kustomization?
> > * Adding seccompProfile, changing unconfined -> runtimeDefault?-
> 
> 1. I would add them to our manifests like you did
> 2. I would add them to the manager as you did - because it's simpler
> 3. I'm fine with adding seccompProfile assuming it will work on all Kubernetes clusters (I'm not really familiar with the seccomp feature in Kubernetes)

Thanks for the reply! We had a discussion in CAPM3 that we want to align with you (CAPI) first on this, and then replicate the approach to CAPM3, BMO, IPAM and elsewhere where necessary. 

As for you question about seccompProfiles, the link in the PR description describes them. `RuntimeDefault` is detailed here: https://kubernetes.io/docs/tutorials/security/seccomp/#create-pod-that-uses-the-container-runtime-default-seccomp-profile 

In short, all (or almost all) container runtimes come with a default profile that limits some of the syscalls from the containers. They're typically not used by non-root workloads, so enabling it changes nothing. In case container calls restricted syscall, it is prevented and app will crash. In our testing, I've enabled it for CAPI, CAPM3, BMO, IPAM without any issues.

> P.S. I assume you wanted to clarify those points first. Eventually we should modify the manager manifests of CABPK, KCP and CAPD in this PR as well (CAPD won't work as non-root because of the docker socket mount, but let's explicitly add what we can)

I can amend the PR for sure. Would you ok-to-test the PR so I can verify CAPI tests pass before I do that?

**Comment by sbueringer** [comment]
/ok-to-test
(sorry missed that)

**Comment by sbueringer** [comment]
> In short, all (or almost all) container runtimes come with a default profile that limits some of the syscalls from the containers. They're typically not used by non-root workloads, so enabling it changes nothing. In case container calls restricted syscall, it is prevented and app will crash. In our testing, I've enabled it for CAPI, CAPM3, BMO, IPAM without any issues.

The interesting question is what happens if the container runtime doesn't come with a default profile.

If I understand it correctly starting with Kubernetes 1.25 the SeccompDefault (https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) feature gate will be enabled per default which sets the default profile automatically.

So I think overall we could have the following problems:
* Does this also work with Kubernetes 1.20 (our minimum supported mgmt cluster Kubernetes version)
* Are there container runtimes that don't have a default profile

Given that starting with Kubernetes 1.25 Kubernetes takes care of setting the default seccomp profile automatically I would prefer not setting it in ClusterAPI do not run into issues in the edge cases. (but just my opinion)

**Comment by tuminoid** [comment]
> > In short, all (or almost all) container runtimes come with a default profile that limits some of the syscalls from the containers. They're typically not used by non-root workloads, so enabling it changes nothing. In case container calls restricted syscall, it is prevented and app will crash. In our testing, I've enabled it for CAPI, CAPM3, BMO, IPAM without any issues.
> 
> The interesting question is what happens if the container runtime doesn't come with a default profile.
> 
> If I understand it correctly starting with Kubernetes 1.25 the SeccompDefault (https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) feature gate will be enabled per default which sets the default profile automatically.

SeccompDefault feature gate enables kubelet to use `--seccomp-default` to enable `RuntimeDefault` for all workloads, but unless it is added to kubelet command line, feature gate is not doing anything. 

`You must also explicitly enable the defaulting behavior for each node where you want to use this with the corresponding --seccomp-default [command line flag](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet). Both have to be enabled simultaneously to use the feature.`

> So I think overall we could have the following problems:
> 
>     * Does this also work with Kubernetes 1.20 (our minimum supported mgmt cluster Kubernetes version)
> 
>     * Are there container runtimes that don't have a default profile
> 
> 
> Given that starting with Kubernetes 1.25 Kubernetes takes care of setting the default seccomp profile automatically I would prefer not setting it in ClusterAPI to not run into issues in the edge cases. (but just my opinion, no objection if folks want to enable it and think it's safe)

- https://kubernetes.io/docs/tutorials/security/seccomp/ seccomp is supported as stable from k8s 1.19
- 1.25 does not have it as default, see above

If the container runtime doesn't have a default profile, then it'll be no different than Unconfined, as the profile filters syscalls, or depending on your runtime configuration, it may already be using runtime defaults.

https://github.com/cri-o/cri-o/blob/main/docs/crio.conf.5.md
- `seccomp_profile="" Path to the seccomp.json profile which is used as the default seccomp profile for the runtime. If not specified, then the internal default seccomp profile will be used.`
- `seccomp_use_default_when_empty=true Changes the meaning of an empty seccomp profile. By default (and according to CRI spec), an empty profile means unconfined. This option tells CRI-O to treat an empty profile as the default profile, which might increase security.`

https://docs.docker.com/engine/security/seccomp/ for Docker. Good list of syscalls filtered, showing most of them are namespaced, and not available anyways as we drop all capabilities, + few obsolete or non-namespaced calls.

containerd follows Docker profile: https://github.com/containerd/containerd/discussions/5924 and https://github.com/containerd/containerd/blob/f0a32c66dad1e9de716c9960af806105d691cd78/contrib/seccomp/seccomp_default.go#L51

and so on.

Of course, if you think seccomp is too risky to enable by default, I can leave it out of this PR.

**Comment by sbueringer** [comment]
Thx for the additional context! Sounds okay to me, but I would like to hear more opinions from others.

**Comment by bengentil** [comment]
Setting a security context in the manifest will break tilt `live_update`

This means for each code update, the pod won't be updated and tilt will report this error:
```
capi_control… │ - '<you local path>/cluster-api/.tiltbuild/bin/manager' --> '/manager'
capi_control… │ tar: can't remove old file manager: Permission denied
```

From what I understood from https://github.com/tilt-dev/tilt/issues/3060,
the only way to make it work right now is to dynamically remove the security context in the Tiltfile.
I can try to propose a fix in that direction if you want.

**Comment by sbueringer** [comment]
@bengentil Nice catch, thx! Given how our tilt setup works I think that is something that we can do in our tilt-prepare binary (roughly [here](https://github.com/kubernetes-sigs/cluster-api/blob/a05d17137d5c981a5ad1d3fb81acaf05e726e426/hack/tools/tilt-prepare/main.go#L769)). But would be good to have verification that dropping it there works.

**Comment by bengentil** [comment]
@sbueringer you're right it's way easier and less error prone to implement it in tilt-prepare

I've successfully tested this fix: https://github.com/bengentil/cluster-api/commit/b3ce57e2b7b15e86683de26214a01dbc41ccb7d2

Don't know the process if it has to be in the same PR, but it should be merged before or at the same time ideally, feel free to take my commit in this branch if needed.

**Comment by sbueringer** [comment]
I think it would be good to have this in this PR to keep tilt working when the PR is merged. Cherry-pick souds fine and it's also okay to keep the tilt change in a separate commit on this PR with you as author.

**Comment by bengentil** [comment]
PR created: https://github.com/kubernetes-sigs/cluster-api/pull/7846

**Comment by tuminoid** [comment]
Thanks @bengentil for spotting the tilt issue and the PR to fix it! It'd be good to have it merged first.

I've updated the PR by splitting the indentation fix to separate commit and by adding securityContext's to KCP and CABPK. CAPD is privileged and none of the securityContexts make sense or have effect when privileged, so I did not touch it.

File: bootstrap/kubeadm/config/manager/manager.yaml [hunk 1]
```diff
@@ -16,31 +16,43 @@ spec:
...
+              runAsUser: 65532
+              runAsGroup: 65532
       terminationGracePeriodSeconds: 10
       serviceAccountName: manager
       tolerations:
         - effect: NoSchedule
           key: node-role.kubernetes.io/master
         - effect: NoSchedule
           key: node-role.kubernetes.io/control-plane
+      securityContext:
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/config/manager/manager.yaml:55]: Q: (for my understanding) do we need a pod security context given that we are overriding it at the container level?

File: bootstrap/kubeadm/config/manager/manager.yaml [hunk 2]
```diff
@@ -16,31 +16,43 @@ spec:
...
+              protocol: TCP
+          readinessProbe:
+            httpGet:
+              path: /readyz
+              port: healthz
+          livenessProbe:
+            httpGet:
+              path: /healthz
+              port: healthz
+            securityContext:
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/config/manager/manager.yaml:40]: Q: (for my understanding) do we need `allowPrivilegeEscalation: false` and `privileged: false` (which AFAIK are the the default for those fields)

**Review by fabriziopandini — COMMENTED** [review]
This is a nice change, I have changed the title so it stands out on release notes!
Can we document this change in upgrade notes for the providers as well?
- https://main.cluster-api.sigs.k8s.io/developer/providers/v1.3-to-v1.4.html#other --> Cluster API controllers are now adding an explicit security context by default.
- https://main.cluster-api.sigs.k8s.io/developer/providers/v1.3-to-v1.4.html#suggested-changes-for-providers --> Providers should add an explicit security context to their controllers, see https://github.com/kubernetes-sigs/cluster-api/pull/7831 for reference

Also, it will be great if we apply security policies to our test extension in https://github.com/kubernetes-sigs/cluster-api/blob/main/test/extension/config/default/manager.yaml (It should be pretty straightforward, but if I'm wrong this could be done also a follow-up PR)

File: bootstrap/kubeadm/config/manager/manager.yaml (hunk 2 above)
Comment (sbueringer) [reply bootstrap/kubeadm/config/manager/manager.yaml:40]: IIRC the securityContexts on pod and container level are not the same (i.e. you can't set the same fields in both)

File: bootstrap/kubeadm/config/manager/manager.yaml (hunk 1 above)
Comment (sbueringer) [reply bootstrap/kubeadm/config/manager/manager.yaml:55]: IIRC the securityContexts on pod and container level are not the same (i.e. you can't set the same fields in both)

**Review by killianmuldoon — COMMENTED** [review]
Looks awesome! Just for safety it would be good to have a full e2e test run to make sure there's nothing conflicting between the security policy and the way CAPI actually runs (I don't think there should be though)

/test pull-cluster-api-e2e-full-main

File: bootstrap/kubeadm/config/manager/manager.yaml (hunk 1 above)
Comment (tuminoid) [reply bootstrap/kubeadm/config/manager/manager.yaml:55]: Indeed, they are setting different settings and on different granularity level, and they are also somewhat relying on each other. For example, deployment level securityContext is setting `runAsNonRoot: true`, which requires container to run as user identifiable by UID (either `USER <UID>` in Dockerfile, and/or `runAsUser: <UID>` in container level securityContext).

File: bootstrap/kubeadm/config/manager/manager.yaml (hunk 2 above)
Comment (tuminoid) [reply bootstrap/kubeadm/config/manager/manager.yaml:40]: What @sbueringer says is correct. `allowPrivilegeEscalation` would require a capability to escalate via sudo, but the point stands.  Privileged container would be effective root, and you wouldn't be able these limitations to privileged container anyways.

Whole point of the PR is to make security configuration explicit, and not rely on defaults or fallbacks. As mentioned in PR description, none of these (besides seccompPolicy, but see earlier comments how it is in-fact quite no-change as well) actually change things, they make it explicit and enforce them.

**Comment by tuminoid** [comment]
> This is a nice change, I have modified the PR title so it stands out on release notes! Can we document this change in upgrade notes for the providers as well?
> 
>     * https://main.cluster-api.sigs.k8s.io/developer/providers/v1.3-to-v1.4.html#other --> Cluster API controllers are now using an explicit security context by default.
> 
>     * https://main.cluster-api.sigs.k8s.io/developer/providers/v1.3-to-v1.4.html#suggested-changes-for-providers --> Providers should add an explicit security context to their controllers deployment, see #7831 for reference
> 
> 
> Also, it will be great if we apply security policies to our test extension in https://github.com/kubernetes-sigs/cluster-api/blob/main/test/extension/config/default/manager.yaml (It should be pretty straightforward, but if I'm wrong this could be done also a follow-up PR)

Thanks for the review, and suggestions! I've added changelog entries as suggested, and also added securitycontext to the test extension. Test extension seems to be built similarly to controllers, so I applied identical securityContext to it.

I'm not too familiar with the test coverage, is the text extension covered by the tests already running, or should some additional test be triggered?

**Comment by fabriziopandini** [comment]
/test pull-cluster-api-e2e-full-main
/lgtm

please squash commits

**Comment by tuminoid** [comment]
> please squash commits

Squashed. e2e passed.

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Comment by tuminoid** [comment]
What's with lint not reporting back?

/test lint

**Comment by killianmuldoon** [comment]
I think lint will have to be retriggered by someone with the correct rights for the repo, but I think we should be happy to merge if an approver verifies that lint is working locally - not sure why it's stuck.

**Comment by tuminoid** [comment]
Rebased due changelog conflict. Please re-review and let's get this merged. This already has had approved/lgtm from everyone required.
